encrypt.go

```diff
@@ -39,47 +39,100 @@ func (es *encryptStream) Write(plaintext []byte) (int, error) {
 	if ret, es.err = es.buffer.Write(plaintext); es.err != nil {
 		return 0, es.err
 	}
-	for es.buffer.Len() >= encryptionBlockSize {
-		es.err = es.encryptBlock()
+
+	// If es.buffer.Len() == encryptionBlockSize, we don't want to
+	// write it out just yet, since for V2 we need to be sure this
+	// isn't the last block.
+	for es.buffer.Len() > encryptionBlockSize {
+		es.err = es.encryptBlock(false)
 		if es.err != nil {
 			return 0, es.err
 		}
 	}
 	return ret, nil
 }
 
-func (es *encryptStream) encryptBlock() error {
-	var n int
-	var err error
-	n, err = es.buffer.Read(es.inblock[:])
-	if err != nil {
-		return err
+func makeEncryptionBlock(version Version, ciphertext []byte, isFinal bool, authenticators []payloadAuthenticator) interface{} {
+	ebV1 := encryptionBlockV1{
+		PayloadCiphertext:  ciphertext,
+		HashAuthenticators: authenticators,
+	}
+	switch version {
+	case Version1():
+		return ebV1
+	case Version2():
+		return encryptionBlockV2{
+			encryptionBlockV1: ebV1,
+			IsFinal:           isFinal,
+		}
+	default:
+		panic(ErrBadVersion{version})
 	}
-	return es.encryptBytes(es.inblock[0:n])
 }
 
-func (es *encryptStream) encryptBytes(b []byte) error {
+func checkEncryptBlockRead(version Version, isFinal bool, blockSize, plaintextLen, bufLen int) {
+	die := func() error {
+		return fmt.Errorf("invalid encryptBlock read state: version=%s, isFinal=%t, plaintextLen=%d, bufLen=%d", plaintextLen, bufLen)
+	}
+
+	// We shouldn't read more than a full block's worth.
+	if plaintextLen > blockSize {
+		die()
+	}
+
+	// If we read less than a full block's worth, then we
+	// shouldn't have anything left in the buffer.
+	if plaintextLen < blockSize && bufLen > 0 {
+		die()
+	}
+
+	switch version {
+	case Version1():
+		// isFinal must be equivalent to plaintextLen being 0
+		// (which, by the above, implies that bufLen == 0).
+		if isFinal != (plaintextLen == 0) {
+			die()
+		}
+
+	case Version2():
+		// If isFinal, then plaintextLen can be any number,
+		// buf bufLen must be 0.
+		if isFinal != (bufLen == 0) {
+			die()
+		}
+
+	default:
+		panic(ErrBadVersion{version})
+	}
+}
+
+func (es *encryptStream) encryptBlock(isFinal bool) error {
+	plaintext := es.buffer.Next(encryptionBlockSize)
```

Comment:
The tradeoff is that we have to be sure not to stash the plaintext slice anywhere, cause it's going to get overwritten later. Even though it's kind of obvious we're not doing that, could we add a comment?